sys/linux/vusb.txt

```diff
@@ -0,0 +1,749 @@
+} [packed]
+
+vusb_descriptor_hid_report {
+	req_type	const[USB_TYPE_STANDARD, int8]
+	desc_type	const[HID_DT_REPORT, int8]
+	len		bytesize[data, int32]
+	data		hid_descriptor_report
+} [packed]
+
+vusb_descriptor_bos {
+	req_type	const[USB_TYPE_STANDARD, int8]
+	desc_type	const[USB_DT_BOS, int8]
+	len		bytesize[data, int32]
+	data		usb_bos_descriptor
+} [packed]
+
+vusb_descriptor_hub_hs {
+	req_type	const[USB_TYPE_STANDARD, int8]
+	desc_type	const[USB_DT_HUB, int8]
+	len		bytesize[data, int32]
+	data		usb_hub_descriptor_hs
+} [packed]
+
+vusb_descriptor_hub_ss {
+	req_type	const[USB_TYPE_STANDARD, int8]
+	desc_type	const[USB_DT_SS_HUB, int8]
+	len		bytesize[data, int32]
+	data		usb_hub_descriptor_ss
+} [packed]
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+vusb_responses {
+	len			len[parent, int32]
+	generic			ptr[in, vusb_response_generic]
+	set_interface		ptr[in, vusb_response_set_interface]
+	get_interface		ptr[in, vusb_response_get_interface]
+	set_configuration	ptr[in, vusb_response_set_configuration]
+	get_configuration	ptr[in, vusb_response_get_configuration]
+	hub_status		ptr[in, vusb_response_get_status_hub]
+	port_status		ptr[in, vusb_response_get_status_port]
+	aiptek_get_report	ptr[in, vusb_response_aiptek_get_report]
+	aiptek_set_report	ptr[in, vusb_response_aiptek_set_report]
+	cdc_ncm_ntb_parameters	ptr[in, vusb_response_cdc_get_ntb_parameters]
+} [packed]
+
+vusb_response_generic {
+	type	flags[usb_request_types, int8]
+	req	flags[usb_requests, int8]
+	len	bytesize[data, int32]
+	data	array[int8]
+} [packed]
+
+usb_requests = USB_REQ_GET_STATUS, USB_REQ_CLEAR_FEATURE, USB_REQ_SET_FEATURE, USB_REQ_SET_ADDRESS, USB_REQ_GET_DESCRIPTOR, USB_REQ_SET_DESCRIPTOR, USB_REQ_GET_CONFIGURATION, USB_REQ_SET_CONFIGURATION, USB_REQ_GET_INTERFACE, USB_REQ_SET_INTERFACE, USB_REQ_SYNCH_FRAME, USB_REQ_SET_SEL, USB_REQ_SET_ISOCH_DELAY, USB_REQ_SET_ENCRYPTION, USB_REQ_GET_ENCRYPTION, USB_REQ_RPIPE_ABORT, USB_REQ_SET_HANDSHAKE, USB_REQ_RPIPE_RESET, USB_REQ_GET_HANDSHAKE, USB_REQ_SET_CONNECTION, USB_REQ_SET_SECURITY_DATA, USB_REQ_GET_SECURITY_DATA, USB_REQ_SET_WUSB_DATA, USB_REQ_LOOPBACK_DATA_WRITE, USB_REQ_LOOPBACK_DATA_READ, USB_REQ_SET_INTERFACE_DS, USB_REQ_GET_PARTNER_PDO, USB_REQ_GET_BATTERY_STATUS, USB_REQ_SET_PDO, USB_REQ_GET_VDM, USB_REQ_SEND_VDM
+
+vusb_response_set_interface {
+	type	const[USB_TYPE_STANDARD, int8]
+	req	const[USB_REQ_SET_INTERFACE, int8]
+	len	const[0, int32]
+} [packed]
+
+vusb_response_get_interface {
+	type	const[USB_TYPE_CLASS, int8]
+	req	const[USB_REQ_GET_INTERFACE, int8]
+	len	bytesize[data, int32]
+	data	int8
+} [packed]
+
+vusb_response_set_configuration {
+	type	const[USB_TYPE_STANDARD, int8]
+	req	const[USB_REQ_SET_CONFIGURATION, int8]
+	len	const[0, int32]
+} [packed]
+
+vusb_response_get_configuration {
```

Comment:
This looks like a good candidate for a template:
```
type vusb_response_t[CLASS, REQ, DATA] {
	type	const[CLASS, int8]
	req	const[REQ, int8]
	len	bytesize[data, int32]
	data	DATA
} [packed]
```

Reply:
Can templates reduce code duplication in other places?

Reply:
Done. There are some USB descriptor descriptions that look a bit similar to each other, but I think it's better to leave them open coded.